barman/cloud_providers/aws_s3.py

```diff
@@ -407,3 +409,37 @@ def _delete_objects_batch(self, paths):
                     % (error_dict["Key"], error_dict["Code"], error_dict["Message"])
                 )
             raise CloudProviderError()
+
+    def get_prefixes(self, prefix):
+        """
+        Return only the common prefixes under the supplied prefix.
+
+        :param str prefix: The object key prefix under which the common prefixes
+            will be found.
+        :rtype: Iterator[str]
+        :return: A list of unique prefixes immediately under the supplied prefix.
+        """
+        for wal_prefix in self.list_bucket(prefix + "/", delimiter="/"):
+            if wal_prefix.endswith("/"):
+                yield wal_prefix
+
+    def delete_under_prefix(self, prefix):
+        """
+        Delete all objects under the specified prefix.
+
+        :param str prefix: The object key prefix under which all objects should be
+            deleted.
+        """
+        if len(prefix) == 0 or prefix == "/" or not prefix.endswith("/"):
```

Comment:
I understand the paranoia, so let's keep this. but how can this happen in the real world?
I had to read the error message twice to really understand it, honestly the problem is not that you are unable to safely remove all the files, the problem is that (rightly) removing a full folder is not allowed.
Maybe we can reword the error?

Reply:
Rewording makes sense - maybe `Deleting all objects under prefix %s is not allowed`?